core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,251 @@
+          HandleMesosEventResponse(
+            response,
+            Set(podId))
+        } else {
+          HandleMesosEventResponse.empty
+        }
+    }
+  }
+
+  /**
+    * We remove a task if it is not reachable and running, and it has no podSpec defined
+    *
+    * Should be called with the effects already applied for the specified podIds
+    * @param podIds podIds changed during the last frame
+    * @return
+    */
+  def pruneTaskStatuses(podStatuses: Map[PodId, PodStatus], podSpecs: Map[PodId, PodSpec])
+                       (podIds: Set[PodId]): ResponseBuilder = {
+    var b = ResponseBuilder.empty
+    def terminalOrUnreachable(status: PodStatus): Boolean = {
+      // Cheesy method
+      status.taskStatuses.values.forall(status => status == Mesos.TaskStatus.TASK_RUNNING)
+    }
+
+    podIds.foreach { podId =>
+      if (podStatuses.contains(podId)) {
+        val podSpecDefined = !podSpecs.contains(podId)
+        // prune terminal statuses for which there's no defined podSpec
+        if (!podSpecDefined && terminalOrUnreachable(podStatuses(podId))) {
+          b = b.withChangeMessage(PodStatusChange(podId, None))
+        }
+      }
+    }
+    b
+  }
+
+  private def podIdFor(taskId: TaskId): PodId = PodId(taskId)
+}
+
+class SchedulerLogic {
+  var podSpecs: Map[PodId, PodSpec] = Map.empty
+  var podRecords: Map[PodId, PodRecord] = Map.empty
+  var podStatuses: Map[PodId, PodStatus] = Map.empty
+
+  var cachedPendingLaunch: Set[PodId] = Set.empty
+
+  private def applyChanges(response: Response): Unit = {
+    response.statusMessages.foreach {
+      case recordChange: PodRecordChange =>
+        recordChange.newRecord match {
+          case Some(newRecord) =>
+            podRecords = podRecords.updated(recordChange.id, newRecord)
+            // podRecord presence means we've launched
+            cachedPendingLaunch -= recordChange.id
+          case None =>
+            podRecords -= recordChange.id
+        }
+      case statusChange: PodStatusChange =>
+        statusChange.newStatus match {
+          case Some(newStatus) =>
+            podStatuses = podStatuses.updated(statusChange.id, newStatus)
+          case None =>
+            podStatuses -= statusChange.id
+        }
+      case agentRecordChange: AgentRecordChange => // TODO
+      case reservationStatusChange: ReservationStatusChange => // TODO
+      case statusSnapshot: StatusSnapshot => // TODO
+    }
+  }
+
+  /**
+    * Maintains the internal cache of pods pending launch, so a full scan isn't required on every offer
+    *
+    * Must be called after podSpecs are updated, and response effects are applied.
+    *
+    * @param podIds
+    */
+  private def updatePendingLaunch(podIds: Set[PodId]): Unit = {
+    podIds.foreach { podId =>
+      val shouldBeLaunched = podSpecs.get(podId).exists { podSpec =>
+        SchedulerLogic.pendingLaunch(podId, podSpec.goal, podRecords.get(podId))
```

Comment:
can we do `import SchedulerLogic._` please? It would make stuff easier to read...